plugins/module_utils/pulp_helper.py

```diff
@@ -51,177 +46,195 @@ def __init__(self, argument_spec={}, **kwargs):
 
     def exit_json(self, changed=False, **kwargs):
         changed |= self._changed
         super(PulpAnsibleModule, self).exit_json(changed=changed, **kwargs)
 
-    def wait_for_task(self, task_href):
-        task = self.tasks_api.read(task_href)
+
+class PulpEntityAnsibleModule(PulpAnsibleModule):
+    def __init__(self, argument_spec={}, **kwargs):
+        spec = dict(
+            state=dict(
+                choices=['present', 'absent'],
+            ),
+        )
+        spec.update(argument_spec)
+        super(PulpEntityAnsibleModule, self).__init__(
+            argument_spec=spec,
+            **kwargs
+        )
+
+
+class PulpEntity(object):
+    def __init__(self, module, natural_key=None, desired_attributes=None):
+        self.module = module
+        self.api_client = self._api_client_class(self.module.api_config)
+        self.api = self._api_class(self.api_client)
+        self.entity = None
+        self.natural_key = natural_key
+        self.desired_attributes = desired_attributes
+
+    def find(self):
+        search_result = self.api.list(**self.natural_key)
+        if search_result.count == 1:
+            self.entity = search_result.results[0]
+            return self.entity
+        else:
+            return None
+
+    def list(self):
+        entities = []
+        offset = 0
+        search_result = self.api.list(limit=PAGE_LIMIT, offset=offset)
+        entities.extend(search_result.results)
+        while search_result.next:
+            offset += PAGE_LIMIT
+            search_result = self.api.list(limit=PAGE_LIMIT, offset=offset)
+            entities.extend(search_result.results)
+        return entities
+
+    def create(self):
+        if not hasattr(self.api, 'create'):
+            self.module.fail_json(msg="This entity is not creatable.")
+        kwargs = dict()
+        kwargs.update(self.natural_key)
+        kwargs.update(self.desired_attributes)
+        entity = self._api_entity_class(**kwargs)
+        if not self.module.check_mode:
+            response = self.api.create(entity)
+            if getattr(response, 'task', None):
+                task = PulpTask(self.module).wait_for(response.task)
+                entity = self.api.read(task.created_resources[0])
+            else:
+                entity = response
+        self.module._changed = True
+        self.entity = entity
+        return self.entity
+
+    def update(self):
+        changed = False
+        desired_attributes = self.desired_attributes.copy()
+        # drop 'file' because artifacts as well as content units are immutable anyway
+        desired_attributes.pop('file', None)
+        for key, value in desired_attributes.items():
+            if getattr(self.entity, key, None) != value:
+                setattr(self.entity, key, value)
+                changed = True
+        if changed:
+            if not hasattr(self.api, 'update'):
+                self.module.fail_json(msg="This entity is immutable.")
```

Comment:
Wouldn't it make sense to check this before changing `self.entity` (so at the entry point of the method like in `create` above)?

Reply:
Actually, this logic isn'r even new...
But it fails only if an actual change is attempted for a reason. Even the module for an immutable entity should be idempotent. If this check was at the beginning, second call with `state=present` would fail instead of succeeding with `changed=false`.